File: zashi-sdk-ios/Sources/Polynomial.swift
import Foundation

internal class Polynomial {
    let coeffs: [Int64]
    let modulus: Int64

    var degree: Int {
        return coeffs.count
    }

    init(coeffs: [Int64], modulus: Int64) {
        self.coeffs = coeffs
        self.modulus = modulus
    }

    static func +(lhs: Polynomial, rhs: Polynomial) -> Polynomial {
        precondition(lhs.degree == rhs.degree, "Polynomial degrees must match")
        precondition(lhs.modulus == rhs.modulus, "Moduli must match")

        let result = (0..<lhs.degree).map { i in
            floorMod(lhs.coeffs[i] + rhs.coeffs[i], lhs.modulus)
        }
        return Polynomial(coeffs: result, modulus: lhs.modulus)
    }

    static func *(lhs: Polynomial, rhs: Polynomial) -> Polynomial {
        precondition(lhs.modulus == rhs.modulus, "Moduli must match")

        let resultDegree = lhs.degree + rhs.degree - 1
        var temp = [Int64](repeating: 0, count: resultDegree)

        for i in 0..<lhs.degree {
            for j in 0..<rhs.degree {
                let product = multiplyMod(lhs.coeffs[i], rhs.coeffs[j], lhs.modulus)
                temp[i + j] = floorMod(temp[i + j] + product, lhs.modulus)
            }
        }

        return Polynomial(coeffs: temp, modulus: lhs.modulus).reduceByXnPlus1(n: lhs.degree)
    }

    private func reduceByXnPlus1(n: Int) -> Polynomial {
        var reduced = [Int64](repeating: 0, count: n)

        for i in 0..<coeffs.count {
            let quotient = i / n
            let remainder = i % n

            if quotient % 2 == 0 {
                reduced[remainder] = Polynomial.floorMod(reduced[remainder] + coeffs[i], modulus)
            } else {
                reduced[remainder] = Polynomial.floorMod(reduced[remainder] - coeffs[i], modulus)
            }
        }

        return Polynomial(coeffs: reduced, modulus: modulus)
    }

    private static func multiplyMod(_ a: Int64, _ b: Int64, _ mod: Int64) -> Int64 {
        let result = (Int(a) * Int(b)) % Int(mod)
        return Int64(result)
    }

    private static func floorMod(_ a: Int64, _ mod: Int64) -> Int64 {
        let result = a % mod
        return result < 0 ? result + mod : result
    }

    func toList() -> [Int64] {
        return coeffs
    }
}

File: zashi-sdk-kotlin/src/main/kotlin/co/electriccoin/zcash/ui/fheanalytics/Polynomial.kt
package co.electriccoin.zcash.ui.fheanalytics

internal class Polynomial(
    val coeffs: LongArray,
    val modulus: Long
) {
    val degree: Int = coeffs.size

    operator fun plus(other: Polynomial): Polynomial {
        require(degree == other.degree) { "Polynomial degrees must match" }
        require(modulus == other.modulus) { "Moduli must match" }

        val result = LongArray(degree) { i ->
            floorMod(coeffs[i] + other.coeffs[i], modulus)
        }
        return Polynomial(result, modulus)
    }

    operator fun times(other: Polynomial): Polynomial {
        require(modulus == other.modulus) { "Moduli must match" }

        val resultDegree = degree + other.degree - 1
        val temp = LongArray(resultDegree)

        for (i in coeffs.indices) {
            for (j in other.coeffs.indices) {
                val product = multiplyMod(coeffs[i], other.coeffs[j], modulus)
                temp[i + j] = floorMod(temp[i + j] + product, modulus)
            }
        }

        return Polynomial(temp, modulus).reduceByXnPlus1(degree)
    }

    private fun reduceByXnPlus1(n: Int): Polynomial {
        val reduced = LongArray(n)

        for (i in coeffs.indices) {
            val quotient = i / n
            val remainder = i % n

            if (quotient % 2 == 0) {
                reduced[remainder] = floorMod(reduced[remainder] + coeffs[i], modulus)
            } else {
                reduced[remainder] = floorMod(reduced[remainder] - coeffs[i], modulus)
            }
        }

        return Polynomial(reduced, modulus)
    }

    private fun multiplyMod(a: Long, b: Long, mod: Long): Long {
        val aBig = a.toBigInteger()
        val bBig = b.toBigInteger()
        val modBig = mod.toBigInteger()
        return ((aBig * bBig) % modBig).toLong()
    }

    private fun floorMod(a: Long, mod: Long): Long {
        val result = a % mod
        return if (result < 0) result + mod else result
    }

    fun toList(): List<Long> = coeffs.toList()

    companion object {
        fun fromDoubleList(values: List<Double>, modulus: Long): Polynomial {
            val coeffs = LongArray(values.size) { i ->
                val rounded = values[i].toLong()
                floorMod(rounded, modulus)
            }
            return Polynomial(coeffs, modulus)
        }

        private fun floorMod(a: Long, mod: Long): Long {
            val result = a % mod
            return if (result < 0) result + mod else result
        }
    }
}
